### DISCUSSION TITLE
Type error occurred when registering route

### DISCUSSION THREAD

**Discussion by myyrakle** [tokio-rs/axum#1576] [Q&A] [answered]
As usual, I tried to register the route by putting the extension in the handler function.
The code for that is:
```
...
pub async fn router() -> Router {
    let app = Router::new().route("/", post(create_room));

    app
}

async fn create_room(
    Json(body): Json<CreateRoomRequest>,
    database: Extension<Arc<Database>>,
    current_user: Extension<Arc<CurrentUser>>,
) -> impl IntoResponse {
    if current_user.authorized == false || current_user.user.is_none() {
        return (StatusCode::UNAUTHORIZED).into_response();
    }
...
```

That code throws the following error:
![image](https://user-images.githubusercontent.com/16988115/204074795-0b10c9e1-26e0-43d9-ba4c-dfa9296c79d0.png)

This is a problem that has not occurred in other codes configured in the same form, so it is difficult to understand.

This is the code point where no error occurred. (https://github.com/tetrust/tetrust-server/blob/develop/src/routes/user/handler.rs#L23)

The code point where the error occurred.
https://github.com/tetrust/tetrust-server/blob/develop/src/routes/room/handler.rs#L23

Can you tell me what am I doing wrong?

**Comment by davidpdrsn** [comment, marked as answer]
Make `Json` the last extractor. See https://docs.rs/axum/latest/axum/extract/index.html#the-order-of-extractors

**Reply by myyrakle** [reply]
Doesn't seem valid to me. If I move the Json to the far right, the result is the same.

```
async fn create_room(
    current_user: Extension<CurrentUser>,
    database: Extension<Arc<Database>>,
    Json(body): Json<CreateRoomRequest>,
) -> impl IntoResponse {
```
![image](https://user-images.githubusercontent.com/16988115/204079089-19b37105-6cfb-4d2e-a870-3650c66cc8b7.png)

### 

I checked while partially deleting the code. If I delete a specific .await statement, it compiles, but if I put it back, an error occurs.
this part.
https://github.com/tetrust/tetrust-server/blob/develop/src/routes/room/handler.rs#L66

I'm a bit confused.

test environment
Windows 11, rustc 1.65. axum 0.5.17

**Reply by davidpdrsn** [reply]
Try https://docs.rs/axum/latest/axum/attr.debug_handler.html

**Reply by myyrakle** [reply]
Found the problem!

It was a Send trait-related error that occurred because there was a function that returned Box<dyn Error>.

Thank you for your kind favor.
